### PR TITLE
Initial release

### PR DESCRIPTION
**Pull request by aapomm**

### CODE REVIEW

File: lib/dradis/plugins/nipper/field_processor.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+module Dradis
+  module Plugins
+    module Nipper
+
+      class FieldProcessor < Dradis::Plugins::Upload::FieldProcessor
+        ALLOWED_DATA_NAMES = %w{section devices}
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/field_processor.rb:6]: alpha order

File: lib/dradis/plugins/nipper/gem_version.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        Gem::Version.new VERSION::STRING
+      end
+
+      module VERSION
+        MAJOR = 3
+        MINOR = 19
+        TINY = 0
+        PRE = nil
+
+        STRING = [MAJOR, MINOR, TINY, PRE].compact.join(".")
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/gem_version.rb:14]:
```suggestion
        STRING = [MAJOR, MINOR, TINY, PRE].compact.join(".")
        STRING = [MAJOR, MINOR, TINY, PRE].compact.join('.')
```

File: lib/nipper/evidence.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+module Nipper
+  class Evidence
+    def initialize(xml_node)
+      @xml = xml_node
+    end
+
+    def supported_tags
+      [ :device_name, :device_type, :device_osversion ]
```
Comment (sean-yeoh) [inline lib/nipper/evidence.rb:8]: alpha order

File: lib/nipper/issue.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    def initialize(xml_node)
+      @xml = xml_node
+    end
+
+    def supported_tags
+      [
+        :title,
+        :cvss_base, :cvss_temporal, :cvss_environmental,
+        :cvss_base_vector, :cvss_temporal_vector, :cvss_environmental_vector,
+        :finding, :impact, :ease, :recommendation
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:12]: alpha order too

File: lib/dradis/plugins/nipper/field_processor.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+        def post_initialize(args={})
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/field_processor.rb:8]:
```suggestion
        def post_initialize(args = {})
```
spacing

File: lib/dradis/plugins/nipper/field_processor.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+          @nipper_object =
+            if data.name == 'section'
+              ::Nipper::Issue.new(data)
+            elsif data.name == 'devices'
+              ::Nipper::Evidence.new(data)
+            end
+        end
+
+        def value(args={})
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/field_processor.rb:19]:
```suggestion
        def value(args = {})
```
spacing

File: lib/dradis/plugins/nipper/importer.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          next if SECURITY_SECTIONS.include?(report.attr('ref'))
+
+          process_issue(report)
+        end
+
+        logger.info { 'Report processed...' }
+      end
+
+      true
+
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/importer.rb:43]:
```suggestion
```

File: lib/dradis/plugins/nipper/importer.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  class Importer < Dradis::Plugins::Upload::Importer
+    SECURITY_SECTIONS = %w{
+      SECURITY.INTRODUCTION
+      SECURITY.CONCLUSIONS
+      SECURITY.RECOMMENDATIONS
+      SECURITY.MITIGATIONS
+      SECURITY.CLASSIFICATIONS
+    }
+
+    def import(params={})
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/importer.rb:11]:
```suggestion
    def import(params = {})
```
spacing

File: lib/dradis/plugins/nipper/field_processor.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
+module Dradis
+  module Plugins
+    module Nipper
+
```
Comment (sean-yeoh) [inline lib/dradis/plugins/nipper/field_processor.rb:4]:
```suggestion
```
remove newline

File: lib/nipper/issue.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      super
+    end
+
+    def method_missing(method, *args)
+      unless supported_tags.include?(method)
+        super
+        return
+      end
+
+      return process_field_value(method)
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:27]:
```suggestion
      process_field_value(method)
```
remove extra return

File: lib/nipper/issue.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        ease: 'section[@ref="EASE"]/text',
+        recommendation: 'section[@ref="RECOMMENDATION"]/text'
+      }
+
+      if method == :title
+        @xml.attr('title')
+      elsif method.to_s.starts_with?('cvss')
+        process_cvss_field(method)
+      else
+        @xml.xpath("./#{translations_table[method].to_s}").first.try(:text)
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:43]:
```suggestion
        @xml.xpath("./#{translations_table[method]}").first.try(:text)
```
redundant `.to_s`

File: lib/nipper/issue.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      translations_table = {
+        cvss_base: 'issuedetails/ratings/cvssv2-base',
+        cvss_temporal: 'issuedetails/ratings/cvssv2-temporal',
+        cvss_environmental: 'issuedetails/ratings/cvssv2-environmental',
+      }
+
+      base_method = method.to_s.sub('_vector', '').to_sym
+
+      if method.to_s.ends_with?('vector')
+        @xml.xpath("./#{translations_table[base_method].to_s}").first.try(:text)
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:57]:
```suggestion
        @xml.xpath("./#{translations_table[base_method]}").first.try(:text)
```
redundant `.to_s`

File: lib/nipper/issue.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+      else
+        @xml.xpath("./#{translations_table[base_method].to_s}").attr('score')
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:59]:
```suggestion
        @xml.xpath("./#{translations_table[base_method]}").attr('score')
```
redundant `.to_s`

File: lib/nipper/issue.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    def supported_tags
+      [
+        :title,
+        :cvss_base, :cvss_temporal, :cvss_environmental,
+        :cvss_base_vector, :cvss_temporal_vector, :cvss_environmental_vector,
+        :finding, :impact, :ease, :recommendation
+      ]
+    end
+
+    def respond_to?(method, include_private=false)
```
Comment (sean-yeoh) [inline lib/nipper/issue.rb:16]:
```suggestion
    def respond_to?(method, include_private = false)
```
spacing

File: lib/nipper/evidence.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    end
+
+    def respond_to?(method, include_private=false)
```
Comment (sean-yeoh) [inline lib/nipper/evidence.rb:11]:
```suggestion
    def respond_to?(method, include_private = false)
```
spacing

File: lib/nipper/evidence.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+    def method_missing(method, *args)
+      unless supported_tags.include?(method)
+        super
+        return
+      end
+
+      field = method.to_s.split('_')[1]
+
+      return @xml.at_xpath('./device').attr(field)
```
Comment (sean-yeoh) [inline lib/nipper/evidence.rb:24]:
```suggestion
      @xml.at_xpath('./device').attr(field)
```
redundant `return`

File: templates/evidence.template [hunk 17]
```diff
@@ -0,0 +1,8 @@
+#[Device Name]#
```
Comment (rachkor) [inline templates/evidence.template:1]:
```suggestion
#[DeviceName]#
```

File: templates/evidence.template [hunk 18]
```diff
@@ -0,0 +1,8 @@
...
+%evidence.device_name%
+
+#[Device Type]#
```
Comment (rachkor) [inline templates/evidence.template:4]:
```suggestion
#[DeviceType]#
```

File: templates/evidence.template [hunk 19]
```diff
@@ -0,0 +1,8 @@
...
+%evidence.device_type%
+
+#[OS Version]#
```
Comment (rachkor) [inline templates/evidence.template:7]:
```suggestion
#[OS]#
```

File: templates/issue.template [hunk 20]
```diff
@@ -0,0 +1,23 @@
+#[Title]#
+%issue.title%
+
+#[CVSS Base]#
```
Comment (rachkor) [inline templates/issue.template:4]:
```suggestion
#[CVSSv2.BaseScore]#
```

File: templates/issue.template [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+%issue.cvss_base%
+
+#[CVSS Temporal]#
```
Comment (rachkor) [inline templates/issue.template:7]:
```suggestion
#[CVSSv2.TemporalScore]#
```

File: templates/issue.template [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+%issue.cvss_temporal%
+
+#[CVSS Environmental]#
```
Comment (rachkor) [inline templates/issue.template:10]:
```suggestion
#[CVSSv2.EnvironmentalScore]#
```

**Review by rachkor — CHANGES_REQUESTED** [review]
I added comments to request a few fields be changed to not have spaces in them. Especially the CVSS score fields! As-is, no one could use the CVSS field names in a filter (you can't filter by "CVSS Score", it has to be spaceless).

File: Gemfile.lock [hunk 23]
```diff
@@ -0,0 +1,103 @@
+PATH
```
Comment (etdsoft) [inline Gemfile.lock:1]: We don't typically include .lock files in gem repos

File: dradis-nipper.gemspec [hunk 24]
```diff
@@ -1,31 +1,28 @@
...
-  spec.license     = "MIT"
+  spec.platform    = Gem::Platform::RUBY
+  spec.name        = 'dradis-nipper'
+  spec.version     = version
+  spec.authors     = ['Daniel Martin']
+  spec.email       = ['etd@nomejortu.com']
+  spec.homepage    = 'http://dradisframework.org'
+  spec.summary     = 'Nipper upload add-on for Dradis Framework.'
+  spec.description = 'This add-on allows you to upload and parse reports from Nipper.'
+  spec.license     = 'GPL-2'
```
Comment (etdsoft) [inline dradis-nipper.gemspec:17]: Can you please use the standard values we use for other gems (see Okta, or Azure)

**Pull request merged**